### ISSUE TITLE
Iteration through graph nodes and set devices

### ISSUE THREAD

**Issue by serjl** [migueldeicaza/TensorFlowSharp#371]
There is a rather simple way to iterate graph nodes and set a device in Java TF:

        String deviceString = String.format("/gpu:%d", deviceIdx);
        GraphDef.Builder builder = GraphDef.parseFrom(graphDef).toBuilder();
        for (int i = 0; i < builder.getNodeCount(); i++) {
            builder.getNodeBuilder(i).setDevice(deviceString);

Is there a corresponding options in the TFC#.

**Comment by hillin** [comment]
I'm also looking for a way to set devices on ALL nodes of an imported graph (due to [this issue](https://github.com/tensorflow/tensorflow/issues/18861)). Python and C++ both have their own way to set device in this scenario.

Python uses `with tf.device...`, which can be easily mimicked in TF#:

```
var graph = new TFGraph();
using (var device = graph.WithDevice(deviceName))
{
     graph.Import(File.ReadAllBytes(path));
}
```
However this does not work, it seems `TFGraph.WithDevice` is just a superficial approach which does not set device for `TFGraph.Import`-ed nodes.

The C++ approach is similar to the Java one @serjl mentioned, iterate through all the nodes and set device for them. But after messing with TFGraph and TFOperation and TFOperationDesc for an hour I realized it's not trivial to do in TF#.

**Comment by serjl** [comment]
I found a way to do it through python. Import your frozen *.pb in python and run over all nodes and rename them all to the desired GPU. Save it again into a new *.pb. Use it in TF# with adding!!!! into sessionOptions the allow_soft_placement=True, so that non-GPU nodes can still run on CPU.

**Comment by AldoZ** [comment]
@serjl do you have sample code you can share please?

**Comment by serjl** [comment]
Later today will put here. But just to stress, it does work one session *.pb per device. So if you want to run on 2 gpus you need to initiate two sessions with its own *.pb uploaded.

**Comment by hillin** [comment]
@serjl yeah I'm aware that you can do it in python, C++ and even Java, but I just want to make my code as C# as possible.

**Comment by serjl** [comment]
@hillin Have  you tried TensorFlow.NET ?

**Comment by serjl** [comment]
> @serjl do you have sample code you can share please? graph.WithDevice does work, but only if the TF# apps are running in separate processes. Your trick of remodelling a frozen pb will simplify things enormously.

detection_graph = tf.Graph()
with detection_graph.as_default():
    od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(PATH_TO_PB, 'rb') as fid:
        serialized_graph = fid.read()
        od_graph_def.ParseFromString(serialized_graph)
        for i in range(2):
            for nod in od_graph_def.node:
                # if nod.name in nodeList:
                    nod.device = "/device:GPU:%d" %i
            tf.train.write_graph(od_graph_def, PATH_TO_NEW_PB, 'frozen_inference_graph_GPU_'+ '%d' % i +'.pb', False)

**Comment by serjl** [comment]
@AldoZ Here the Python code for two GPU's:
detection_graph = tf.Graph()
with detection_graph.as_default():
    od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(PATH_TO_PB, 'rb') as fid:
        serialized_graph = fid.read()
        od_graph_def.ParseFromString(serialized_graph)
        for i in range(2):
            for nod in od_graph_def.node:
                # if nod.name in nodeList:
                    nod.device = "/device:GPU:%d" %i
            tf.train.write_graph(od_graph_def, PATH_TO_NEW_PB, 'frozen_inference_graph_GPU_'+ '%d' % i +'.pb', False)

**Comment by AldoZ** [comment]
@serjl Fantastic, thank you

**Comment by hillin** [comment]
Here is my approach (as I said, as C# as possible):

```
[DllImport("libtensorflow")]
private static extern void TF_ImportGraphDefOptionsSetDefaultDevice(IntPtr opts, string device);

private static void SetDevice(TFImportGraphDefOptions options, string device)
{
    TF_ImportGraphDefOptionsSetDefaultDevice(options.Handle, device);
}

private static TFGraph LoadGraph(string path, string deviceName)
{
    var graph = new TFGraph();

    var options = new TFImportGraphDefOptions();
    SetDevice(options, deviceName);

    graph.Import(File.ReadAllBytes(path), options);

    return graph;
}

public TFSession RunSession(string graphPath, string deviceName)
{
    var graph = LoadGraph(path, deviceName);

    // follow this guide to use ConfigProto: https://github.com/migueldeicaza/TensorFlowSharp/issues/206
    var config = new ConfigProto
    {
        LogDevicePlacement = true,
        AllowSoftPlacement = true   
    };

    using (var stream = new MemoryStream())
    {
        config.WriteTo(stream);
        var configBuffer = stream.ToArray();

        var options = new TFSessionOptions();

        unsafe
        {
            fixed (byte* pConfigBuffer = configBuffer)
            {
                options.SetConfig((IntPtr)pConfigBuffer, configBuffer.Length);
            }
        }
        
        return new TFSession(graphPath, options);
    }
}
```

So `RunSession(graphPath, "/device:GPU:1"` gives you a session which will run on GPU1.

**Comment by serjl** [comment]
@hillin Cool, thank you very much! Regarding the return in CreateSession: shouln't it be 
`return new TFSession(graph, options); `
otherwise I don't see where you use the graph variable.

**Comment by hillin** [comment]
@serjl yeh it was a typo. Fixed!
